pkg/cli/command.go

```diff
@@ -0,0 +1,145 @@
+package cli
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"strings"
+
+	"github.com/posener/complete"
+	"github.com/spf13/pflag"
+)
+
+// Command represents a (sub)command of the application. Either `Run()` must be
+// defined, or subcommands added using `AddCommand()`. These are also mutually
+// exclusive.
+type Command struct {
+	// Usage line. First word must be the command name, everything else is
+	// displayed as-is.
+	Use string
+	// Aliases define alternative names for a command
+	Aliases []string
+
+	// Short help text, used for overviews
+	Short string
+	// Long help text, used for full help pages. `Short` is used as a fallback
+	// if unset.
+	Long string
+
+	// Version of the application. Only used on the root command
+	Version string
+
+	// Run is the action that is run when this command is invoked.
+	// The error is returned as-is from `Execute()`.
+	Run func(cmd *Command, args []string) error
+
+	// Validation + Completion
+	//
+	// Predict contains Predictors for flags. Defaults to
+	// `complete.PredictSomething` if unset.
+	// Use the flags name (not shorthand) as the key.
+	Predictors map[string]complete.Predictor
+	// Args is used to validate and complete positional arguments
+	Args Arguments
+
+	// internal fields
+	children  []*Command
+	flags     *pflag.FlagSet
+	parentPtr *Command
+}
+
+// Execute runs the application. It should be run on the most outer level
+// command.
+// The error return value is used for both, application errors but also help texts.
+func (c *Command) Execute() error {
+	// Execute must be called on the top level command
+	if c.parentPtr != nil {
+		return c.parentPtr.Execute()
+	}
+
+	// add subcommand for install CLI completions
+	c.AddCommand(completionCmd(c.Use))
+
+	// add version flag to root command. If `findTarget` switches to a
+	// subcommand afterwards, showVersion will stay nil, causing the version not
+	// to be shown.
+	var showVersion *bool
+	if c.Version != "" {
+		showVersion = c.Flags().Bool("version", false, fmt.Sprintf("version for %s", c.Use))
+	}
+
+	// exit if in bash completion context
+	if predict(c) {
+		return nil
+	}
+
+	// find the correct (sub)command
+	c, args, err := findTarget(c, os.Args[1:])
```

Comment:
It threw me off a little that we're shadowing `c` here, could we give the subcommand variable another name?

Reply:
I know that pain, it annoyed me all the time. Not shadowing it also.

The thing here is that the first command needs to find the command that should actually be executed .. 

Once we have that there is no point in keeping the initial command around, using it can lead to unnoticed and hard-to-debug issues. That's why I decided to shadow it

Reply:
If we leave it as shadowed, maybe instead the comment at the bottom saying we're going to run notes that we're running the discovered subcommand?